### PR TITLE
Bokeh 0.11 Compatibility

### PR DESCRIPTION
**Pull request by philippjfr**
This PR adds various fixes to ensure HoloViews is compatible with the new 0.11 bokeh release.

It is now ready for merge.

### CODE REVIEW

File: holoviews/plotting/bokeh/widgets.py [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+from distutils.version import LooseVersion
+
 import param
+import bokeh
+from bokeh.io import push_notebook, Document
+
+if LooseVersion(bokeh.__version__) >= LooseVersion('0.11'):
+    old_bokeh = False
+    from bokeh.io import _CommsHandle
+    from bokeh.util.notebook import get_comms
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/widgets.py]: How much can we trust that ``_CommsHandle`` is something we can rely on in future? From the name, it doesn't look like something we are expected to use directly...

File: holoviews/plotting/bokeh/renderer.py [hunk 2]
```diff
@@ -56,23 +59,9 @@ def __call__(self, obj, fmt=None):
...
-            json_data = []
-            for plotobj in plotobjects:
-                if plotobj.ref['id'] in ids:
-                    continue
-                else:
-                    ids.append(plotobj.ref['id'])
-                if old_bokeh:
-                    json = plotobj.vm_serialize(changed_only=True)
-                else:
-                    json = plotobj.to_json(False)
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/renderer.py]: If you find this pattern happening often, I would consider using a utility what does the version check and smooths over the differences between bokeh versions. I'll let you decide if this this worth it - if you only need it in this one instance, it probably doesn't need to be made into a utility right now.

File: holoviews/plotting/bokeh/widgets.py (hunk 1 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/widgets.py]: Certainly possible that they will change it but there's no way around it. Bokeh notebook support is currently set up to only keep one CommsHandle open at the same so we can't rely on their core API. It's certainly worth keeping in mind though and replace it with a call to a higher-level API once/if that becomes possible.

File: holoviews/plotting/bokeh/callbacks.py [hunk 3]
```diff
@@ -9,12 +9,13 @@
     old_bokeh = True
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/callbacks.py]: ``old_bokeh`` isn't a great name as one day the latest version will become old! I would consider stating the version explicitly...i.e ``bokeh_XX``.

**Comment by jlstevens** [comment]
I've made some comments above but I'm afraid I don't know enough details to do a very thorough review. I do have one general comment though...

The changes here seem to be of two types 1. actual changes in the bokeh API to smooth over 2. improvements to how the code works that is compatible across both bokeh versions (i.e understanding how to leverage bokeh better).

The second type of change seems general and I assume fairly stable. The first type of change is a lot of special cases and I worry that as bokeh keeps developing (and the API keeps getting tweaked) that the list of these fixes will keep growing. I'm not sure it is needed *now* but I might consider a class in util or even a file ``compat.py`` if we think the number of version dependent fixes will keep expanding in future.

File: holoviews/plotting/bokeh/widgets.py (hunk 1 above)
Comment (jbednar) [inline holoviews/plotting/bokeh/widgets.py]: It's worth posting an issue on the Bokeh site asking whether this is an ok long-term solution, and/or request a better one be made available.

File: holoviews/plotting/bokeh/callbacks.py (hunk 3 above)
Comment (jbednar) [inline holoviews/plotting/bokeh/callbacks.py]: +1

File: holoviews/plotting/bokeh/widgets.py (hunk 1 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/widgets.py]: I know that handling multiple Comms at once is on their radar and might even land in 0.11.1. If need be I can mirror CommsHandle in HoloViews, it's a tiny class that's literally just meant to hold a few attributes.

File: holoviews/plotting/bokeh/renderer.py (hunk 2 above)
Comment (philippjfr) [inline holoviews/plotting/bokeh/renderer.py]: I think I ever repeat the same pattern multiple times.

Comment (jbednar) [inline holoviews/plotting/bokeh/renderer.py]: even, or never?

**Comment by philippjfr** [comment]
Agree that we'll have to add a compat.py module if there are a lot more changes to the API coming. It could simply supply various version agnostic functions to do specific things. Currently it doesn't seem worth it though.

**Comment by jlstevens** [comment]
Tests are all passing so I'll merge now.

**Pull request merged**